### DISCUSSION TITLE
Is there a way to turn off anti-aliasing for font drawing, or some other way to make fonts pixelated but not blurred?

### DISCUSSION THREAD

**Discussion by tbell5** [hajimehoshi/ebiten#1941] [Q&A] [answered]
Hello,

I'm looking for the functionality to turn off anti-aliasing when drawing fonts, a functionality useful for rendering 8-bit style fonts from standard TrueType fonts, and which appears, for instance, in raster drawing applications like GIMP. See for example the attached text sample, which I rendered in GIMP using the Georgia font with anti-aliasing turned off, which can be scaled up to create a pixelated aesthetic.

![text_sample](https://user-images.githubusercontent.com/43974313/148660427-448475d4-8b7f-4b48-8d8e-4f4d9cb557c4.png)

I only recently started learning about the implementation of fonts in Go (or in general, for that matter), but I've seen no mention of anti-aliasing in either the Go [font package](https://pkg.go.dev/golang.org/x/image/font) or the Ebiten [text package](https://pkg.go.dev/github.com/hajimehoshi/ebiten/text), and little discussion on the matter. My best guess would be that it just isn't implemented in Go, but I wanted to make sure I wasn't missing anything.

If turning off anti-aliasing isn't an option in Go or Ebiten, could anyone offer any guidance as to how I should go about rendering pixelated fonts? The whole area is confusing to me, and I haven't had any luck finding a solution over the course of a few days or researching the topic, though I'll keep looking. I am aware of @hajimehoshi's [bitmapfont package](https://pkg.go.dev/github.com/hajimehoshi/bitmapfont), which could work decently well for my use case, but I would still prefer a way to render any given font in the way specified above.

Thanks!

**Comment by eihigh** [comment, marked as answer]
Because the font is converted from vector format to image by "rasterization", I don't think it would be easy to properly convert it to a 8-bit style bitmap.

As far as I can tell, as you guessed, you're going to have to use bitmap fonts (Hajime's bitmapfont is one of them).

**Reply by tbell5** [reply]
I follow what you're saying, but I think it's a matter of how the rasetrizer... rasterizes (?). In the GIMP example, for instance, the drawing application is still taking the font's TTF (vector) specification and mapping what bits to draw through rasterization. The only difference is that when you turn off anti-aliasing in GIMP, the rasterizer doesn't shade any of the pixels in semi-shades of grey to try and "smooth out" the text. So this is all I'm looking for, a way to make the rasterizer (implemented by the [opentype package](https://pkg.go.dev/golang.org/x/image@v0.0.0-20211028202545-6944b10bf410/font/opentype) for TTF fonts in Go) behave like it does when you have anti-aliasing turned off in GIMP.

**Reply by eihigh** [reply]
OK, I understand, the font package in Go doesn't provide an API for that, so it looks like you'll need to modify it, but I don't know how...

**Reply by tbell5** [reply]
Yea, I think you're right. The [wikipedia article on font rasterization](https://en.wikipedia.org/wiki/Font_rasterization) shows a good example of rasterization with and without anti-aliasing, which makes me think it would be a good feature to include in the API, or at least in a forked version or something. Definitely over my head as well, though...
